testcases.py

```diff
@@ -692,6 +692,87 @@ def check(self) -> TestResult:
         return TestResult.SUCCEEDED
 
 
+class TestCaseAmplificationLimit(TestCase):
+    @staticmethod
+    def name():
+        return "amplificationlimit"
+
+    @staticmethod
+    def testname(p: Perspective):
+        return "transfer"
+
+    @staticmethod
+    def abbreviation():
+        return "A"
+
+    @staticmethod
+    def desc():
+        return "The server obeys the 3x amplification limit."
+
+    def certs_dir(self):
+        if not self._cert_dir:
+            self._cert_dir = tempfile.TemporaryDirectory(dir="/tmp", prefix="certs_")
+            generate_cert_chain(self._cert_dir.name, 9)
+        return self._cert_dir.name + "/"
+
+    @staticmethod
+    def scenario() -> str:
+        """ Scenario for the ns3 simulator """
+        # Let the ClientHello pass, but drop a bunch of retransmissions afterwards.
+        return "droplist --delay=15ms --bandwidth=10Mbps --queue=25 --drops_to_server=2,3,4,5,6,7"
+
+    def get_paths(self):
+        self._files = [self._generate_random_file(5 * KB)]
+        return self._files
+
+    def check(self) -> TestResult:
+        if not self._keylog_file():
+            logging.info("Can't check test result. SSLKEYLOG required.")
+            return TestResult.UNSUPPORTED
+        num_handshakes = self._count_handshakes()
+        if num_handshakes != 1:
+            logging.info("Expected exactly 1 handshake. Got: %d", num_handshakes)
+            return TestResult.FAILED
+        if not self._check_version_and_files():
+            return TestResult.FAILED
+        # Check the highest offset of CRYPTO frames sent by the server.
+        # This way we can make sure that it actually used the provided cert chain.
+        max_handshake_offset = 0
+        for p in self._server_trace().get_handshake(Direction.FROM_SERVER):
+            if hasattr(p, "crypto_offset"):
+                max_handshake_offset = max(
+                    max_handshake_offset, int(p.crypto_offset) + int(p.crypto_length)
+                )
+        if max_handshake_offset < 7500:
```

Comment:
Where does this number come from? Is the cert chain guaranteed to be larger than 7500? Could you do a character count of all files in the certs dir (`"wc -c " + self._cert_dir.name + "/* | grep total | tr -s ' ' | cut -f2 -d' '"`) and use that instead?

Reply:
Agreed, this is a magic number. But to be precise, we'd have to include the TLS framing overhead for the Certificate message, and somehow quantify the possible saving by using certificate compression (which people are free to use, if they want to), so we'll end up with something hand-wavy anyway.